### PR TITLE
8253284: Zero OrderAccess barrier mappings are incorrect

### PR DESCRIPTION
**Pull request by shipilev**
There are some jcstress failures with AArch64 Zero. It seems because to happen because `orderAccess_linux_zero.hpp` defaults to compiler-only barriers for most OrderAccess::* calls. We need to defer to the strongest barriers by default. 

The code also needs some rearrangement to make the mappings clear.

jcstress seems to capture the bug, and seems to pass when bug is fixed. Since this is `zero`, we want only the interpreter (compiler configs would be excessive). Release builds capture more samples.

```
$ wget https://builds.shipilev.net/jcstress/jcstress-tests-all-20200917.jar
$ build/linux-x86_64-server-release/images/jdk/bin/java -jar jcstress-tests-all-20200917.jar --jvmArgs "-Xint" 
```

Testing:
 - [x] x86_64 Linux zero release jcstress run
 - [x] x86_64 MacOS zero release jcstress run
 - [x] AArch64 Linux zero release jcstress run
 - [x] PPC Linux zero release jcstress run
 - [x] SKIPPED: ARM32 Linux zero release jcstress run (it would seem ARM32 is broken already: JDK-8253464)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253284](https://bugs.openjdk.java.net/browse/JDK-8253284): Zero OrderAccess barrier mappings are incorrect

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)
 * [Andrew John Hughes](https://openjdk.java.net/census#andrew) (@gnu-andrew - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/224/head:pull/224`
`$ git checkout pull/224`

### CODE REVIEW

File: src/hotspot/os_cpu/bsd_zero/orderAccess_bsd_zero.hpp [hunk 1]
```diff
@@ -28,8 +28,6 @@
 
 // Included in orderAccess.hpp header file.
 
-#ifdef ARM
```
Comment (dholmes-ora) [inline src/hotspot/os_cpu/bsd_zero/orderAccess_bsd_zero.hpp:31]: I assume this should have been changed to #if defined(ARM) ??

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Looks reasonable other than the ifdef ARM issue.

File: src/hotspot/os_cpu/bsd_zero/orderAccess_bsd_zero.hpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/os_cpu/bsd_zero/orderAccess_bsd_zero.hpp:31]: Dang! Right, copy-paste error. Fixed.

**Comment by gnu-andrew** [comment]
Is the #if defined(ARM) referring to ARM in general here, or just 32-bit ARM? Because there appears to be no change to its definition, just rearrangement.
The patch is quite hard to follow, and made even more difficult by multiple versions being attached to this PR.

**Comment by shipilev** [comment]
> Is the #if defined(ARM) referring to ARM in general here, or just 32-bit ARM? Because there appears to be no change to its definition, just rearrangement.

No, there is `ARM` (32) and then there is `AARCH64` (64). This is why AArch64 fails: it effectively uses compiler barriers only through the confusing application of #else branches.

> The patch is quite hard to follow, and made even more difficult by multiple versions being attached to this PR.

These are not multiple versions, those commits would be squashed into one on push -- that is how Skara workflow works. Look at the final version, which basically enables default (implicitly used by `ALPHA` and `AARCH64`) strong barriers -- that is the bug fix. It also keeps the light-weight barrier light for `X86` by calling it out explicitly.

**Review by theRealAph — APPROVED** [review]

**Comment by gnu-andrew** [comment]
> > Is the #if defined(ARM) referring to ARM in general here, or just 32-bit ARM? Because there appears to be no change to its definition, just rearrangement.
> 
> No, there is `ARM` (32) and then there is `AARCH64` (64). This is why AArch64 fails: it effectively uses compiler barriers only through the confusing application of #else branches.

Ok, patch looks ok then. That seems to differ from the other architectures, where 'PPC' and 'X86' are referring to both 32- and 64-bit variants.

> 
> > The patch is quite hard to follow, and made even more difficult by multiple versions being attached to this PR.
> 
> These are not multiple versions, those commits would be squashed into one on push -- that is how Skara workflow works. Look at the final version, which basically enables default (implicitly used by `ALPHA` and `AARCH64`) strong barriers -- that is the bug fix. It also keeps the light-weight barrier light for `X86` by calling it out explicitly.

If there are not multiple versions, how is there a 'final version'? :-)

I eventually found the webrev with the full and final patch. One would normally rebase the patch with amendments, not pile commit on commit and then leave it to a bot to rebase. I don't like the idea that the final commit will look different to what is being reviewed.

/reviewer credit andrew

**Review by gnu-andrew — APPROVED** [review]

**Comment by gnu-andrew** [comment]
Why is there this /reviewer command that doesn't work?

**Comment by shipilev** [comment]
> Why is there this /reviewer command that doesn't work?

Reviewer command is to tell the bot who are the off-GH reviewers are. In retrospect, I could have just mentioned you myself here...

**Comment by dholmes-ora** [comment]
The final version of this is the "sum" of the applied commits. When viewing the "changed files" you can select which commit(s) to look at. There is no need to rebase (nor is it even desirable) as you will lose the context of review comments.

Final update looks fine to me.

**Review by dholmes-ora — APPROVED** [review]
